{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1c9ad6be",
   "metadata": {},
   "source": [
    "# Week 8\n",
    "\n",
    "Hopefully you have already opened this document as a Jupyter Notebook and not simply as a pdf. If not, then you need to download the file to an appropriate folder, and open the file.\n",
    "\n",
    "To open the file, if you have already installed Anaconda Individual Edition, then open Anaconda.Navigator, from which you can launch Jupyter Notebook.\n",
    "\n",
    "(If working in the School of Computing labs, start Anaconda3 / Anaconda Prompt, then enter the following command specifying the path to the folder containing the file, and then open the file from the browser.)\n",
    "\n",
    "The Jupyter Notebook App is a tool for creating documents (notebooks) containing both live code and text, as well as visualizations, etc. Various programming languages can be used, but here the focus will be on Python.\n",
    "\n",
    "For more general resources:\n",
    "\n",
    "Jupyter Notebook - https://jupyter.org/\n",
    "Anaconda - https://www.anaconda.com/\n",
    "Python - https://www.python.org/\n",
    "The rest of this notebook is intended as an introduction to key features of Python and Jupyter Notebooks, irrespective of whether you have used Python before or not. It is not intended to be exhaustive, but will cover material that will be relevant in this module. In addition to this material, you are strongly encouarged to develop your knowledge of Python further through the use of other resources such as\n",
    "\n",
    "The Python tutorial - https://docs.python.org/3/tutorial/\n",
    "Both Python and Jupyter Notebooks are used in various areas, including data science, so it would be well worth your while to develop your skills in this area as much as possible.\n",
    "\n",
    "Getting started\n",
    "One thing you should do at the outset is to save this noteback as a new notebook, so that you can change it as much as you want, but still go back to the original file if necessary. Go to File / Make a Copy. This creates a new notebook called Week_1_Lab-Copy1 in a file with the same name. You can change the name of the notebook (and the file) by going to File / Rename and changing it to Week_1_Lab_my_version for example, or by editing the name of the notebook at the top of the page beside the Jupyter symbol (just above the menu).\n",
    "\n",
    "Cell Types\n",
    "We need to distinguish between different types of cells. This cell is a Markdown cell, whereas the next cell below is a Code cell. Markdown cells are for formatting text rather than for running code. To see how to format headings, use italics and bold font, for example, you can go into edit mode by double-clicking on a cell. Try it for this cell. To execute the cell (and so produce the formatted text), you can go to Cell / Run Cells or use the shortcut Ctrl-Enter. (You can find other keyboard shortcuts under the Help menu.) Markdown is very useful for mathematical notation such as  2⎯⎯√\n",
    " . For further details on Markdown see Markdown in Jupyter Notebook.\n",
    "\n",
    "The next cell below is a Code cell. You can also edit and run it as described above, but now it will execute the code and present the output below the cell."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6a37ca4",
   "metadata": {},
   "source": [
    "# Task 1\n",
    "\n",
    "Practice Problem: Python Function with if, for, and while\n",
    "\n",
    "Write a function called process_numbers that takes a list of numbers as input and performs the following steps:\n",
    "\n",
    "\n",
    "If the list is empty, return the string: \"The list is empty\".\n",
    "\n",
    "Use a for loop to iterate through the list:\n",
    "\n",
    "    If a number is even, add it to a new list called even_numbers.\n",
    "    If a number is odd, add it to a new list called odd_numbers.\n",
    "    If the number 0 appears in the list, print \"Zero found in the list\".\n",
    "\n",
    "Use a while loop to calculate the sum of numbers in even_numbers.\n",
    "\n",
    "Return a dictionary with:\n",
    "\n",
    "    even_numbers: the list of even numbers.\n",
    "    odd_numbers: the list of odd numbers.\n",
    "    even_sum: the sum of the even numbers.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b01218a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Zero found in the list\n",
      "{'even_numbers': [10, 20, 0, 8], 'odd_numbers': [15, 3, 7], 'even_sum': 38}\n"
     ]
    }
   ],
   "source": [
    "def process_numbers(numbers):\n",
    "    if not numbers:\n",
    "        return \"The list is empty\"\n",
    "    \n",
    "    even_numbers = []\n",
    "    odd_numbers = []\n",
    "    even_sum = 0\n",
    "    \n",
    "    # Use a for loop to classify even and odd numbers\n",
    "    for num in numbers:\n",
    "        if num % 2 == 0:\n",
    "            even_numbers.append(num)\n",
    "        else:\n",
    "            odd_numbers.append(num)\n",
    "        if num == 0:\n",
    "            print(\"Zero found in the list\")\n",
    "    \n",
    "    # Use a while loop to calculate the sum of even numbers\n",
    "    index = 0\n",
    "    while index < len(even_numbers):\n",
    "        even_sum += even_numbers[index]\n",
    "        index += 1\n",
    "    \n",
    "    # Return the results as a dictionary\n",
    "    return {\n",
    "        \"even_numbers\": even_numbers,\n",
    "        \"odd_numbers\": odd_numbers,\n",
    "        \"even_sum\": even_sum\n",
    "    }\n",
    "\n",
    "# Test\n",
    "numbers = [10, 15, 20, 0, 3, 8, 7]\n",
    "result = process_numbers(numbers)\n",
    "print(result)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07fe0d48",
   "metadata": {},
   "source": [
    "Zero found in the list\n",
    "{'even_numbers': [10, 20, 0, 8], 'odd_numbers': [15, 3, 7], 'even_sum': 38}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a35097ad",
   "metadata": {},
   "source": [
    "## Task 2\n",
    "\n",
    "#### Objective: Write a Python function to analyze DNA sequences for specific characteristics.\n",
    "\n",
    "##### Problem:\n",
    "You are tasked with creating a Python program that:\n",
    "\n",
    "    Checks whether a DNA sequence is valid (contains only A, T, C, and G).\n",
    "    Counts the occurrences of each nucleotide in the sequence.\n",
    "    Finds the complementary sequence using a loop.\n",
    "    Stops processing when a stop codon (TAA, TAG, or TGA) is detected.\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "eed630aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter the DNA sequence: AAAAAGGGGGTCA\n",
      "Enter the stop codon (e.g., TAA, TAG, TGA): TCA\n",
      "{'Nucleotide Count': {'A': 6, 'T': 1, 'C': 1, 'G': 5}, 'Complementary Sequence': 'TTTTTCCCCCAGT', 'Processed Sequence': 'AAAAAGGGGGTCA'}\n"
     ]
    }
   ],
   "source": [
    "def analyze_dna(sequence, stop_codon):\n",
    "    \"\"\"\n",
    "    Analyze a DNA sequence with a user-defined stop codon.\n",
    "\n",
    "    Parameters:\n",
    "        sequence (str): The DNA sequence to analyze.\n",
    "        stop_codon (str): The stop codon provided by the user.\n",
    "\n",
    "    Returns:\n",
    "        dict: A dictionary containing analysis results.\n",
    "    \"\"\"\n",
    "    # Validate the DNA sequence\n",
    "    for base in sequence:\n",
    "        if base not in \"ATCG\":\n",
    "            return {\"Error\": \"Invalid DNA sequence.\"}\n",
    "\n",
    "    # Count occurrences of each nucleotide\n",
    "    nucleotide_count = {\"A\": 0, \"T\": 0, \"C\": 0, \"G\": 0}\n",
    "    for base in sequence:\n",
    "        if base in nucleotide_count:\n",
    "            nucleotide_count[base] += 1\n",
    "\n",
    "    # Find the complementary sequence\n",
    "    complement = {\"A\": \"T\", \"T\": \"A\", \"C\": \"G\", \"G\": \"C\"}\n",
    "    complementary_sequence = \"\"\n",
    "    for base in sequence:\n",
    "        complementary_sequence += complement[base]\n",
    "\n",
    "    # Stop processing at the user-defined stop codon\n",
    "    stop_index = -1\n",
    "    i = 0\n",
    "    while i <= len(sequence) - 3:\n",
    "        codon = sequence[i:i+3]\n",
    "        if codon == stop_codon:\n",
    "            stop_index = i\n",
    "            break\n",
    "        i += 3\n",
    "\n",
    "    if stop_index != -1:\n",
    "        sequence = sequence[:stop_index]  # Truncate at the stop codon\n",
    "\n",
    "    # Return the analysis results\n",
    "    return {\n",
    "        \"Nucleotide Count\": nucleotide_count,\n",
    "        \"Complementary Sequence\": complementary_sequence,\n",
    "        \"Processed Sequence\": sequence\n",
    "    }\n",
    "\n",
    "# Get inputs from the user\n",
    "sequence = input(\"Enter the DNA sequence: \").upper()\n",
    "stop_codon = input(\"Enter the stop codon (e.g., TAA, TAG, TGA): \").upper()\n",
    "\n",
    "# Validate the stop codon\n",
    "if len(stop_codon) != 3 or any(base not in \"ATCG\" for base in stop_codon):\n",
    "    print(\"Invalid stop codon. Please enter a valid codon (3 bases: A, T, C, G).\")\n",
    "else:\n",
    "    # Analyze the DNA sequence\n",
    "    results = analyze_dna(sequence, stop_codon)\n",
    "    print(results)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40e5a571",
   "metadata": {},
   "source": [
    "## Task 3\n",
    "\n",
    "Problem: DNA Sequence Analysis\n",
    "Write a Python script that includes the following functionality:\n",
    "\n",
    "    1. A function called validate_dna(sequence) that checks if a given string is a valid DNA sequence. A valid DNA sequence contains only the characters A, T, C, and G. If invalid, the function should return False.\n",
    "\n",
    "    2. A function called gc_content(sequence) that calculates the GC content (percentage of G and C bases) in a given DNA sequence.\n",
    "\n",
    "    3. A function called find_motif(sequence, motif) that uses a for loop to find all start positions of a given motif (substring) in the DNA sequence.\n",
    "\n",
    "    4. A function called reverse_complement(sequence) that calculates the reverse complement of a DNA sequence using a while loop."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1af995a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GC Content: 50.00%\n",
      "Motif 'CGT' found at positions: [3, 7]\n",
      "Reverse Complement: AGCTAGCTACGTACGCAT\n"
     ]
    }
   ],
   "source": [
    "# 1. Validate DNA sequence\n",
    "def validate_dna(sequence):\n",
    "    \"\"\"\n",
    "    Validate if the sequence contains only A, T, C, G.\n",
    "    Returns True if valid, otherwise False.\n",
    "    \"\"\"\n",
    "    valid_bases = {'A', 'T', 'C', 'G'}\n",
    "    for base in sequence:\n",
    "        if base not in valid_bases:\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "# 2. Calculate GC content\n",
    "def gc_content(sequence):\n",
    "    \"\"\"\n",
    "    Calculate the GC content in a DNA sequence.\n",
    "    Returns GC content as a percentage.\n",
    "    \"\"\"\n",
    "    gc_count = 0\n",
    "    for base in sequence:\n",
    "        if base in {'G', 'C'}:\n",
    "            gc_count += 1\n",
    "    return (gc_count / len(sequence)) * 100\n",
    "\n",
    "# 3. Find all occurrences of a motif\n",
    "def find_motif(sequence, motif):\n",
    "    \"\"\"\n",
    "    Find all start positions of a motif in the DNA sequence.\n",
    "    Returns a list of starting indices.\n",
    "    \"\"\"\n",
    "    positions = []\n",
    "    for i in range(len(sequence) - len(motif) + 1):\n",
    "        if sequence[i:i + len(motif)] == motif:\n",
    "            positions.append(i)\n",
    "    return positions\n",
    "\n",
    "# 4. Reverse complement of DNA\n",
    "def reverse_complement(sequence):\n",
    "    \"\"\"\n",
    "    Calculate the reverse complement of a DNA sequence.\n",
    "    Uses a while loop to process the sequence.\n",
    "    \"\"\"\n",
    "    complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}\n",
    "    reverse_comp = []\n",
    "    i = len(sequence) - 1\n",
    "    while i >= 0:\n",
    "        reverse_comp.append(complement[sequence[i]])\n",
    "        i -= 1\n",
    "    return ''.join(reverse_comp)\n",
    "\n",
    "# Example usage\n",
    "dna_sequence = \"ATGCGTACGTAGCTAGCT\"\n",
    "motif = \"CGT\"\n",
    "\n",
    "if validate_dna(dna_sequence):\n",
    "    print(f\"GC Content: {gc_content(dna_sequence):.2f}%\")\n",
    "    print(f\"Motif '{motif}' found at positions: {find_motif(dna_sequence, motif)}\")\n",
    "    print(f\"Reverse Complement: {reverse_complement(dna_sequence)}\")\n",
    "else:\n",
    "    print(\"Invalid DNA sequence!\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "069207b7",
   "metadata": {},
   "source": [
    "## Advance task\n",
    "\n",
    "The following Python code is intended to calculate the GC content of a given DNA sequence. However, it contains several bugs. Your task is to debug the code and ensure it works correctly.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d9f20ebc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50.0\n",
      "50.0\n",
      "50.0\n",
      "Error: DNA sequence is empty\n",
      "Error: Invalid character in DNA sequence\n",
      "Error: DNA sequence must be a string\n"
     ]
    }
   ],
   "source": [
    "def calculate_gc_content(dna_sequence):\n",
    "    if not isinstance(dna_sequence, str):\n",
    "        return \"Error: DNA sequence must be a string\"\n",
    "    \n",
    "    # Convert the sequence to uppercase to handle lowercase input\n",
    "    dna_sequence = dna_sequence.upper()\n",
    "    \n",
    "    gc_count = 0\n",
    "    total_count = len(dna_sequence)\n",
    "    \n",
    "    if total_count == 0:  # Handle empty sequence case\n",
    "        return \"Error: DNA sequence is empty\"\n",
    "    \n",
    "    for base in dna_sequence:\n",
    "        if base == \"G\" or base == \"C\":\n",
    "            gc_count += 1  # Corrected: Added assignment operator\n",
    "        elif base not in [\"A\", \"T\", \"G\", \"C\"]:\n",
    "            return \"Error: Invalid character in DNA sequence\"\n",
    "    \n",
    "    # Removed unnecessary while loop\n",
    "    gc_content = (gc_count / total_count) * 100\n",
    "    return gc_content\n",
    "\n",
    "# Test Cases\n",
    "print(calculate_gc_content(\"ATGC\"))           # Expected: 50.0\n",
    "print(calculate_gc_content(\"atgc\"))           # Expected: 50.0\n",
    "print(calculate_gc_content(\"AGGCGTAA\"))       # Expected: 50.0\n",
    "print(calculate_gc_content(\"\"))               # Expected: Error: DNA sequence is empty\n",
    "print(calculate_gc_content(\"ATGXYZ\"))         # Expected: Error: Invalid character in DNA sequence\n",
    "print(calculate_gc_content(12345))            # Expected: Error: DNA sequence must be a string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36e14a2c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
