{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1c9ad6be",
   "metadata": {},
   "source": [
    "# Week 7\n",
    "\n",
    "Hopefully you have already opened this document as a Jupyter Notebook and not simply as a pdf. If not, then you need to download the file to an appropriate folder, and open the file.\n",
    "\n",
    "To open the file, if you have already installed Anaconda Individual Edition, then open Anaconda.Navigator, from which you can launch Jupyter Notebook.\n",
    "\n",
    "(If working in the School of Computing labs, start Anaconda3 / Anaconda Prompt, then enter the following command specifying the path to the folder containing the file, and then open the file from the browser.)\n",
    "\n",
    "The Jupyter Notebook App is a tool for creating documents (notebooks) containing both live code and text, as well as visualizations, etc. Various programming languages can be used, but here the focus will be on Python.\n",
    "\n",
    "For more general resources:\n",
    "\n",
    "Jupyter Notebook - https://jupyter.org/\n",
    "Anaconda - https://www.anaconda.com/\n",
    "Python - https://www.python.org/\n",
    "The rest of this notebook is intended as an introduction to key features of Python and Jupyter Notebooks, irrespective of whether you have used Python before or not. It is not intended to be exhaustive, but will cover material that will be relevant in this module. In addition to this material, you are strongly encouarged to develop your knowledge of Python further through the use of other resources such as\n",
    "\n",
    "The Python tutorial - https://docs.python.org/3/tutorial/\n",
    "Both Python and Jupyter Notebooks are used in various areas, including data science, so it would be well worth your while to develop your skills in this area as much as possible.\n",
    "\n",
    "Getting started\n",
    "One thing you should do at the outset is to save this noteback as a new notebook, so that you can change it as much as you want, but still go back to the original file if necessary. Go to File / Make a Copy. This creates a new notebook called Week_1_Lab-Copy1 in a file with the same name. You can change the name of the notebook (and the file) by going to File / Rename and changing it to Week_1_Lab_my_version for example, or by editing the name of the notebook at the top of the page beside the Jupyter symbol (just above the menu).\n",
    "\n",
    "Cell Types\n",
    "We need to distinguish between different types of cells. This cell is a Markdown cell, whereas the next cell below is a Code cell. Markdown cells are for formatting text rather than for running code. To see how to format headings, use italics and bold font, for example, you can go into edit mode by double-clicking on a cell. Try it for this cell. To execute the cell (and so produce the formatted text), you can go to Cell / Run Cells or use the shortcut Ctrl-Enter. (You can find other keyboard shortcuts under the Help menu.) Markdown is very useful for mathematical notation such as  2⎯⎯√\n",
    " . For further details on Markdown see Markdown in Jupyter Notebook.\n",
    "\n",
    "The next cell below is a Code cell. You can also edit and run it as described above, but now it will execute the code and present the output below the cell."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6a37ca4",
   "metadata": {},
   "source": [
    "# Task 1\n",
    "\n",
    "Generates a random DNA sequence and calculates its GC content, which is the percentage of guanine (G) or cytosine (C) bases.\n",
    "\n",
    "Once an acceptable sequence is found (with GC content above 50%), it stops and prints the final sequence. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b01218a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generated sequence: GATACGGGCTCCAGTCGGTGCCGAGGCTAGTATGAGAATCCATACGTACTACATCGACGGGTGCGCCCAGACACAGCGATGTGCGACGCGTAGATGTAGG with GC content: 58.00%\n",
      "\n",
      "Acceptable sequence found: GATACGGGCTCCAGTCGGTGCCGAGGCTAGTATGAGAATCCATACGTACTACATCGACGGGTGCGCCCAGACACAGCGATGTGCGACGCGTAGATGTAGG with GC content: 58.00%\n"
     ]
    }
   ],
   "source": [
    "import random\n",
    "\n",
    "def generate_random_dna_sequence(length):\n",
    "    \"\"\"Generates a random DNA sequence of a given length.\"\"\"\n",
    "    return ''.join(random.choices('ACGT', k=length))\n",
    "\n",
    "def calculate_gc_content(dna_sequence):\n",
    "    \"\"\"Calculates the GC content of a DNA sequence.\"\"\"\n",
    "    gc_count = dna_sequence.count('G') + dna_sequence.count('C')\n",
    "    return (gc_count / len(dna_sequence)) * 100\n",
    "\n",
    "# Set the length of the DNA sequence and the GC content threshold\n",
    "sequence_length = 100\n",
    "gc_threshold = 50  # 50%\n",
    "\n",
    "# Initialize variables\n",
    "gc_content = 0\n",
    "dna_sequence = ''\n",
    "\n",
    "# Use a while loop to keep generating sequences until the GC content is above the threshold\n",
    "while gc_content < gc_threshold:\n",
    "    dna_sequence = generate_random_dna_sequence(sequence_length)\n",
    "    gc_content = calculate_gc_content(dna_sequence)\n",
    "    print(f\"Generated sequence: {dna_sequence} with GC content: {gc_content:.2f}%\")\n",
    "\n",
    "# Print the final acceptable sequence and its GC content\n",
    "print(f\"\\nAcceptable sequence found: {dna_sequence} with GC content: {gc_content:.2f}%\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca891b17",
   "metadata": {},
   "source": [
    "# Task 2\n",
    "\n",
    "Parsing genetic sequence data.\n",
    "\n",
    "1. Parse through a dictionary of gene sequences.\n",
    "2. Calculate the GC content, which is the percentage of bases in a DNA sequence that are either guanine (G) or cytosine (C).\n",
    "3. Search for a specific DNA motif within the sequence.\n",
    "4. Create the reverse complement of the DNA sequence.\n",
    "5. Use control structures (if, for, while) efficiently to accomplish the tasks."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ec305157",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene: gene1\n",
      "  GC Content: 52.17%\n",
      "  Motif 'ATG' Present: Yes\n",
      "  Reverse Complement: CAGTCAGCTAGCTACGTACGCAT\n",
      "\n",
      "Gene: gene2\n",
      "  GC Content: 56.52%\n",
      "  Motif 'ATG' Present: Yes\n",
      "  Reverse Complement: CAGCTAGCATGCTAGCATAGCGC\n",
      "\n",
      "Gene: gene3\n",
      "  GC Content: 50.00%\n",
      "  Motif 'ATG' Present: Yes\n",
      "  Reverse Complement: CGATCGTACGATATGCGCAT\n",
      "\n",
      "First gene with GC content over 50%: gene1\n"
     ]
    }
   ],
   "source": [
    "# Define a dictionary with gene identifiers as keys and DNA sequences as values\n",
    "gene_sequences = {\n",
    "    'gene1': 'ATGCGTACGTAGCTAGCTGACTG',\n",
    "    'gene2': 'GCGCTATGCTAGCATGCTAGCTG',\n",
    "    'gene3': 'ATGCGCATATCGTACGATCG',\n",
    "    # Add more gene sequences as needed\n",
    "}\n",
    "\n",
    "# Function to calculate GC content\n",
    "def calculate_gc_content(sequence):\n",
    "    gc_count = sequence.count('G') + sequence.count('C')\n",
    "    return (gc_count / len(sequence)) * 100\n",
    "\n",
    "# Function to search for a DNA motif\n",
    "def search_motif(sequence, motif='ATG'):\n",
    "    return motif in sequence\n",
    "\n",
    "# Function to create the reverse complement of a DNA sequence\n",
    "def reverse_complement(sequence):\n",
    "    complement = {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G'}\n",
    "    return ''.join(complement[base] for base in reversed(sequence))\n",
    "\n",
    "# Main processing loop\n",
    "for gene_id, sequence in gene_sequences.items():\n",
    "    gc_content = calculate_gc_content(sequence)\n",
    "    motif_present = search_motif(sequence)\n",
    "    rev_complement = reverse_complement(sequence)\n",
    "    \n",
    "    # Print the results for each gene\n",
    "    print(f\"Gene: {gene_id}\")\n",
    "    print(f\"  GC Content: {gc_content:.2f}%\")\n",
    "    print(f\"  Motif 'ATG' Present: {'Yes' if motif_present else 'No'}\")\n",
    "    print(f\"  Reverse Complement: {rev_complement}\\n\")\n",
    "\n",
    "    # Additional conditional checks can be added here\n",
    "    # For example, to process genes with high GC content differently\n",
    "    if gc_content > 60:\n",
    "        print(f\"  Note: {gene_id} has a high GC content.\\n\")\n",
    "\n",
    "# While loop example: Let's say we want to find the first gene with GC content over 50%\n",
    "# and we want to do it using a while loop for demonstration purposes\n",
    "index = 0\n",
    "gene_keys = list(gene_sequences.keys())\n",
    "while index < len(gene_keys):\n",
    "    gene_id = gene_keys[index]\n",
    "    sequence = gene_sequences[gene_id]\n",
    "    if calculate_gc_content(sequence) > 50:\n",
    "        print(f\"First gene with GC content over 50%: {gene_id}\")\n",
    "        break\n",
    "    index += 1\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d7168b6",
   "metadata": {},
   "source": [
    "# Task 3 \n",
    "\n",
    "Exercise: You're working with a dataset of genetic sequences where each sequence is represented by a string composed of four nucleotides (adenine (A), cytosine (C), guanine (G), and thymine (T)). You want to perform the following tasks:\n",
    "\n",
    "1. Create a dictionary where each key is a unique genetic sequence (string) and its value is a tuple containing the count of each nucleotide in the sequence.\n",
    "2. For each sequence, determine if it has a high GC content. GC content is considered high if it is greater than 60% of the sequence.\n",
    "3. Create a set that contains only the sequences with high GC content.\n",
    "4. Write a loop that iteratively reduces the sequences by removing one nucleotide from the end of each sequence until the length of each sequence is less than 10 or no sequences are left."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "709d1ebb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nucleotide counts for each sequence: {'ATCGGCTA': (2, 2, 2, 2), 'GCGCGCGC': (0, 4, 4, 0), 'ATATATAT': (4, 0, 0, 4), 'CGCGATCGA': (2, 3, 3, 1)}\n",
      "High GC content sequences: {'CGCGATCGA', 'GCGCGCGC'}\n",
      "Reduced sequences: []\n"
     ]
    }
   ],
   "source": [
    "# Define a function to calculate GC content\n",
    "def calculate_gc_content(sequence):\n",
    "    gc_content = (sequence.count('G') + sequence.count('C')) / len(sequence)\n",
    "    return gc_content > 0.6  # Returns True if GC content is higher than 60%\n",
    "\n",
    "# Example list of genetic sequences\n",
    "sequences = ['ATCGGCTA', 'GCGCGCGC', 'ATATATAT', 'CGCGATCGA']\n",
    "\n",
    "# Step 1: Create a dictionary with counts of each nucleotide\n",
    "sequence_info = {}\n",
    "for seq in sequences:\n",
    "    # Count each nucleotide and store in a tuple\n",
    "    a_count = seq.count('A')\n",
    "    c_count = seq.count('C')\n",
    "    g_count = seq.count('G')\n",
    "    t_count = seq.count('T')\n",
    "    sequence_info[seq] = (a_count, c_count, g_count, t_count)\n",
    "\n",
    "# Step 2 and 3: Determine sequences with high GC content and add to a set\n",
    "high_gc_content_sequences = set()\n",
    "for seq, counts in sequence_info.items():\n",
    "    if calculate_gc_content(seq):\n",
    "        high_gc_content_sequences.add(seq)\n",
    "\n",
    "# Step 4: Iteratively reduce sequences until they are less than 10 nucleotides long\n",
    "while sequences:  # While the list of sequences is not empty\n",
    "    sequences = [seq[:-1] for seq in sequences if len(seq) > 1]  # Reduce sequences\n",
    "    # Remove sequences that are less than 10 nucleotides long\n",
    "    sequences = [seq for seq in sequences if len(seq) >= 10]\n",
    "\n",
    "# Print the results\n",
    "print(\"Nucleotide counts for each sequence:\", sequence_info)\n",
    "print(\"High GC content sequences:\", high_gc_content_sequences)\n",
    "print(\"Reduced sequences:\", sequences)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2fbd065",
   "metadata": {},
   "source": [
    "# Task 4\n",
    "\n",
    "Calculates the frequency of each nucleotide in a DNA sequence and identifies the most frequent nucleotide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2efd8b27",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nucleotide counts: {'A': 20, 'C': 12, 'G': 17, 'T': 21}\n",
      "Sorted nucleotide counts: [('T', 21), ('A', 20), ('G', 17), ('C', 12)]\n",
      "Most frequent nucleotides: {'T'}\n"
     ]
    }
   ],
   "source": [
    "# Define a DNA sequence (in practice, this would be read from a file)\n",
    "dna_sequence = \"AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGC\"\n",
    "\n",
    "# Initialize a dictionary to store nucleotide counts\n",
    "nucleotide_counts = {\"A\": 0, \"C\": 0, \"G\": 0, \"T\": 0}\n",
    "\n",
    "# Use a for loop to iterate through each nucleotide in the sequence\n",
    "for nucleotide in dna_sequence:\n",
    "    # Check if the nucleotide is one of the expected bases\n",
    "    if nucleotide in nucleotide_counts:\n",
    "        # If so, increment the count for this nucleotide in the dictionary\n",
    "        nucleotide_counts[nucleotide] += 1\n",
    "\n",
    "# Initialize a list to store nucleotides and their counts as tuples\n",
    "nucleotide_count_list = []\n",
    "\n",
    "# Use a for loop to iterate over the dictionary items\n",
    "for nucleotide, count in nucleotide_counts.items():\n",
    "    # Append a tuple of nucleotide and its count to the list\n",
    "    nucleotide_count_list.append((nucleotide, count))\n",
    "\n",
    "# Sort the list of tuples by the count in descending order\n",
    "nucleotide_count_list.sort(key=lambda x: x[1], reverse=True)\n",
    "\n",
    "# Initialize a set to keep track of the most frequent nucleotides\n",
    "most_frequent_nucleotides = set()\n",
    "\n",
    "# Use a while loop to find all nucleotides with the highest frequency\n",
    "# We use the first element's count as the highest count for comparison\n",
    "highest_count = nucleotide_count_list[0][1]\n",
    "index = 0\n",
    "while index < len(nucleotide_count_list) and nucleotide_count_list[index][1] == highest_count:\n",
    "    # Add the nucleotide to the set of most frequent nucleotides\n",
    "    most_frequent_nucleotides.add(nucleotide_count_list[index][0])\n",
    "    index += 1\n",
    "\n",
    "# Output the results\n",
    "print(\"Nucleotide counts:\", nucleotide_counts)\n",
    "print(\"Sorted nucleotide counts:\", nucleotide_count_list)\n",
    "print(\"Most frequent nucleotides:\", most_frequent_nucleotides)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75ad99f7",
   "metadata": {},
   "source": [
    "# Advance Task 5\n",
    "\n",
    "## Exercise Description\n",
    "Given a DNA sequence, perform the following:\n",
    "\n",
    "#### 1. Transcribe the DNA sequence into RNA.\n",
    "\n",
    "Ensure that only valid DNA nucleotides are transcribed.\n",
    "   \n",
    "   \n",
    "#### 2. Translate the RNA sequence into a peptide chain.\n",
    "\n",
    "Create a dictionary representing the codon table.\n",
    "Use a for loop to iterate through the RNA sequence in codons (triplets) and map each to a peptide.\n",
    "   \n",
    "#### 3. Metabolize specific peptides in the sequence into metabolites.\n",
    "\n",
    "If a specific peptide sequence is encountered, break it down into a set of metabolites.\n",
    "Store the metabolites in a list, and ensure no duplicates by converting it to a set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "36029b23",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RNA Sequence: UACCGGUAACAUUACCCGGCGACUUUCCCACGGGCUAUC\n",
      "Peptide Chain: ['Tyrosine', 'Arginine', 'Methionine']\n",
      "Metabolites: {'metabolite2', 'metabolite1'}\n"
     ]
    }
   ],
   "source": [
    "# Define the codon table as a dictionary mapping RNA codons to peptides\n",
    "codon_table = {\n",
    "    'AUG': 'Methionine', 'UGG': 'Tryptophan', 'UAC': 'Tyrosine', 'CGG': 'Arginine', # Example codons\n",
    "    # ... (other codons should be filled in here for a complete table)\n",
    "    # 'UAA', 'UAG', 'UGA' are stop codons and are not included in the codon table\n",
    "}\n",
    "\n",
    "# Transcription function: DNA -> RNA\n",
    "def transcribe_dna_to_rna(dna_sequence):\n",
    "    transcription_map = {'A': 'U', 'T': 'A', 'C': 'G', 'G': 'C'}\n",
    "    rna_sequence = ''\n",
    "    for nucleotide in dna_sequence:\n",
    "        if nucleotide in transcription_map:\n",
    "            rna_sequence += transcription_map[nucleotide]\n",
    "        else:\n",
    "            # Handle invalid nucleotides\n",
    "            raise ValueError(f\"Invalid DNA nucleotide: {nucleotide}\")\n",
    "    return rna_sequence\n",
    "\n",
    "# Translation function: RNA -> Peptide Chain\n",
    "def translate_rna_to_peptide(rna_sequence):\n",
    "    peptide_chain = []\n",
    "    stop_codons = {'UAA', 'UAG', 'UGA'}  # Common stop codons\n",
    "\n",
    "    # Ensure that the RNA sequence length is a multiple of 3\n",
    "    if len(rna_sequence) % 3 != 0:\n",
    "        raise ValueError(\"RNA sequence length is not a multiple of 3\")\n",
    "    \n",
    "    for i in range(0, len(rna_sequence), 3):\n",
    "        codon = rna_sequence[i:i+3]\n",
    "        # Check if codon is a stop codon\n",
    "        if codon in stop_codons:\n",
    "            break  # Stop translation\n",
    "        if codon in codon_table:\n",
    "            peptide_chain.append(codon_table[codon])\n",
    "        else:\n",
    "            # Handle invalid codons\n",
    "            raise ValueError(f\"Invalid RNA codon: {codon}\")\n",
    "    \n",
    "    return peptide_chain\n",
    "\n",
    "# Metabolization function: Peptides -> Metabolites\n",
    "def metabolize_peptides(peptide_chain):\n",
    "    metabolizing_peptides = ('Methionine', 'Tryptophan')\n",
    "    metabolites = []\n",
    "    i = 0\n",
    "    while i < len(peptide_chain):\n",
    "        if peptide_chain[i] in metabolizing_peptides:\n",
    "            metabolites += ['metabolite1', 'metabolite2']\n",
    "            i += 1  # Corrected incrementation\n",
    "        else:\n",
    "            i += 1  # Increment i for all peptides\n",
    "    return set(metabolites)\n",
    "\n",
    "# Main function to run the simulation with a forced metabolization for demonstration\n",
    "def simulate(dna_sequence):\n",
    "    rna_sequence = transcribe_dna_to_rna(dna_sequence)\n",
    "    peptide_chain = translate_rna_to_peptide(rna_sequence)\n",
    "    \n",
    "    # Force the addition of 'Methionine' to the peptide chain for demonstration\n",
    "    peptide_chain.append('Methionine')\n",
    "    \n",
    "    metabolites = metabolize_peptides(peptide_chain)\n",
    "    return rna_sequence, peptide_chain, metabolites\n",
    "\n",
    "# Print the results\n",
    "    dna_seq = 'ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG'\n",
    "    rna_seq, peptides, metabolites = simulate(dna_seq)\n",
    "    print(f\"RNA Sequence: {rna_seq}\")\n",
    "    print(f\"Peptide Chain: {peptides}\")\n",
    "    print(f\"Metabolites: {metabolites}\")\n",
    "except ValueError as e:\n",
    "    print(f\"An error occurred: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77d8de9e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
